### PR TITLE
DEV: Reads in and parses data from SRW dat file 

### PR DESCRIPTION
**Pull request by kalebswartz7**
Uses data from DAT file to create a 2D image based on the points. Parameters for the data are put into a dictionary for easy access. 

Link to beamline: https://beta.sirepo.com/srw#/beamline/F5wzzwl9?coherence=full

Attn: @mrakitin

### CODE REVIEW

File: kalebswartz7/excercise_01/parse.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# Parse the data from the dat file -> Points vs Horizontal Position = how many rows per column in the 2D array 
+#                                  -> Points vs Vertical Position = how many columns in the 2D array 
+
+#       Create a dictionary and populate it with other dictionaries based on the data in the dat file 
+#       Use this data to add to the graph
+
+import matplotlib as mpl
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Why do you need this?

File: kalebswartz7/excercise_01/parse.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+#Just default values, these are changed based on the dat file
+horizontal = 1 
+vertical = 1
+
+#initialize the dictionary for storing parameters given in the dat file 
+initialData = {}
+initialData['photon'] = {}
+initialData['horizontal'] = {}
+initialData['vertical'] = {}
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: You can use a shorter version, e.g.:
```py
initial_data = {
    'photon': {},
    'horizontal': {},
    'vertical': {},
}
```

File: kalebswartz7/excercise_01/parse.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+matrix = [[]] 
+
+""" 
+Get data from the dat file with the parameters for the scan. Put this data into an organized Dictionary
+
+"""
+def getCommentedData():
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Please use [PEP8-standard](https://www.python.org/dev/peps/pep-0008/), i.e. `get_commented_data(...)`.

File: kalebswartz7/excercise_01/parse.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    with open('res_int_pr_se.dat') as file:
+        lineNum = 1
+        for line in file:
+            if "#" not in line:
+                return
+            else:
+                if (lineNum is 1):
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: We don't need `()` with `if`statements. Same for `for`, `while` loops, etc.

Comment (mrakitin) [reply kalebswartz7/excercise_01/parse.py]: It's more natural to use `==` with numbers.

File: kalebswartz7/excercise_01/parse.py (hunk 3 above)
Comment (mrakitin) [reply kalebswartz7/excercise_01/parse.py]: Please also take care of the rest of functions and variables names.

File: kalebswartz7/excercise_01/parse.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+            lineNum += 1
+
+"""
+Funciton that populates the dictionary based on what values from the line are needed, what type they should be, and if they need to change 
+the values of any global varaiables 
+
+"""
+
+def populateDictionary(line, startingPoint, endingPoint, d, d1 = {}, type1='string', vToChange = 0):
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: You don't need blank spaces in the signature of the function, so it would be `d1={}`. Or better use `d1=None`.

Comment (mrakitin) [reply kalebswartz7/excercise_01/parse.py]: Change `vToChange` to `True` or `False` or `None`.

File: kalebswartz7/excercise_01/parse.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                elif ('vs Horizontal' in line):
+                    populateDictionary(line, '#', ' ', 'horizontal', 'Points', 'float','horizontal')
+                elif ('Initial Vertical' in line):
+                    populateDictionary(line, '#', ' ', 'vertical', 'Initial_Position')
+                elif ('Final Vertical' in line):
+                    populateDictionary(line, '#', ' ', 'vertical', 'Final_Position')
+                elif ('vs Vertical' in line):
+                    populateDictionary(line, '#', ' ', 'vertical', 'Points', 'float','vertical')
+
+
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: It's not recommended to have 2 blank lines within a function/class. They are normally used to separate different classes/functions.

File: kalebswartz7/excercise_01/parse.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    elif (changeV):
+        setVertical(vToChange)
+
+def setHorizontal(value):
+    global horizontal
+    horizontal = value
+
+def setVertical(value):
+    global vertical
+    vertical = value
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Better to do it using passing an argument to corresponding plotting functions. The vertical and horizontal number of points can be used from the returned dictionary. Better to not use globals, but pass values as args to functions and return as needed.

File: kalebswartz7/excercise_01/parse.py (hunk 5 above)
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Also, please add a docstring. See an example at https://github.com/NSLS-II/bluesky/blob/868c42f6b6be33f320bb3b5100d59c5378d2b5af/bluesky/plans.py#L25-L46.

File: kalebswartz7/excercise_01/parse.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+matrix = [[]] 
+
+""" 
+Get data from the dat file with the parameters for the scan. Put this data into an organized Dictionary
+
+"""
+def getCommentedData():
+    with open('res_int_pr_se.dat') as file:
+        lineNum = 1
+        for line in file:
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: You can use `contents = file.readlines()` for small files (a few MB), or `file.readline()` with a loop of known length (10-20 header lines).

Comment (mrakitin) [reply kalebswartz7/excercise_01/parse.py]: Please use this as a reference:
```py
In [24]: content = []
    ...: with open('res_int_pr_se.dat', 'r') as f:
    ...:     for i in range(20):
    ...:         content.append(f.readline())
    ...:

In [25]:

In [25]: content = [x.strip() for x in content]

In [26]: content
Out[26]:
['# [ph/s/.1%bw/mm^2] (C-aligned, inner loop is vs Photon Energy, outer loop vs Vertical Position)',
 '#9000.0 #Initial Photon Energy [eV]',
 '#9000.0 #Final Photon Energy [eV]',
 '#1 #Number of points vs Photon Energy',
 '#-0.00014237188284963375 #Initial Horizontal Position [m]',
 '#0.00014256428526195097 #Final Horizontal Position [m]',
 '#528 #Number of points vs Horizontal Position',
 '#-6.997475022319695e-05 #Initial Vertical Position [m]',
 '#6.99747502231969e-05 #Final Vertical Position [m]',
 '#294 #Number of points vs Vertical Position',
 '88871.125',
 '90252.6640625',
 '94020.2265625',
 '96684.3203125',
 '96943.7421875',
 '98506.265625',
 '102780.25',
 '106489.9453125',
 '108295.4140625',
 '109468.7734375']

In [27]: content = [x.strip() for x in content if x.startswith('#')]

In [28]: content
Out[28]:
['# [ph/s/.1%bw/mm^2] (C-aligned, inner loop is vs Photon Energy, outer loop vs Vertical Position)',
 '#9000.0 #Initial Photon Energy [eV]',
 '#9000.0 #Final Photon Energy [eV]',
 '#1 #Number of points vs Photon Energy',
 '#-0.00014237188284963375 #Initial Horizontal Position [m]',
 '#0.00014256428526195097 #Final Horizontal Position [m]',
 '#528 #Number of points vs Horizontal Position',
 '#-6.997475022319695e-05 #Initial Vertical Position [m]',
 '#6.99747502231969e-05 #Final Vertical Position [m]',
 '#294 #Number of points vs Vertical Position']

```

File: kalebswartz7/excercise_01/parse.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                elif ('Initial Photon' in line):
+                    populateDictionary(line, '#', ' ', 'photon', 'Initial_Energy')
+                elif ('Final Photon' in line):
+                    populateDictionary(line, '#', ' ', 'photon', 'Final_Energy')
+                elif ('Initial Horizontal' in line):
+                    populateDictionary(line, '#', ' ', 'horizontal', 'Initial_Position')
+                elif ('Final Horizontal' in line):
+                    populateDictionary(line, '#', ' ', 'horizontal', 'Final_Position')
+                elif ('vs Horizontal' in line):
+                    populateDictionary(line, '#', ' ', 'horizontal', 'Points', 'float','horizontal')
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: It's worth to use the keyword arguments here to easily understand what it does.

File: kalebswartz7/excercise_01/parse.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                    v += 1
+
+def populateMatrixSmart():
+    d = np.loadtxt('res_int_pr_se.dat')
+    d2 = d.reshape((294, 528))
+    global matrix
+    matrix = d2
+
+def createImage(): 
+   im = plt.imshow(matrix, cmap='gray', aspect="equal")
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Please use the same style of quotes everywhere. Either is fine (`'` or `"`).

File: kalebswartz7/excercise_01/parse.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+#       Create a dictionary and populate it with other dictionaries based on the data in the dat file 
+#       Use this data to add to the graph
+
+import matplotlib as mpl
+import numpy as np
+import matplotlib.pyplot as plt
+
+#Just default values, these are changed based on the dat file
+horizontal = 1 
+vertical = 1
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Let's get rid of globals.

File: kalebswartz7/excercise_01/parse.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+                matrix[v][h] = float(line)
+                if (h < horizontal - 1):
+                    h += 1
+                else:
+                    h = 0
+                    v += 1
+
+def populateMatrixSmart():
+    d = np.loadtxt('res_int_pr_se.dat')
+    d2 = d.reshape((294, 528))
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: Make those numbers coming from the parsed file.

**Review by mrakitin — CHANGES_REQUESTED** [review]
Please implement my suggestions. Also, you can apply [`flake8 <script>`](http://flake8.pycqa.org/en/latest/) to see the problems with styles/formatting.

File: kalebswartz7/excercise_01/parse.py (hunk 1 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: It is now gone.

File: kalebswartz7/excercise_01/parse.py (hunk 2 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: Implemented.

File: kalebswartz7/excercise_01/parse.py (hunk 3 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: They have been adjusted.

File: kalebswartz7/excercise_01/parse.py (hunk 4 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: These have been changed.

File: kalebswartz7/excercise_01/parse.py (hunk 5 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: This has been fixed.

File: kalebswartz7/excercise_01/parse.py (hunk 6 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: This has also been fixed.

File: kalebswartz7/excercise_01/parse.py [hunk 13]
```diff
@@ -13,91 +12,86 @@
 vertical = 1
 
 #initialize the dictionary for storing parameters given in the dat file 
-initialData = {}
-initialData['photon'] = {}
-initialData['horizontal'] = {}
-initialData['vertical'] = {}
+initial_data = {'photon': {}, 'horizontal': {}, 'vertical': {}}
```
Comment (mrakitin) [inline kalebswartz7/excercise_01/parse.py]: 👍

**Comment by mrakitin** [comment]
@kalebswartz7, that's the code formatter I mentioned today (thanks to @tacaswell for introducing it to me): https://github.com/ambv/black.

File: kalebswartz7/excercise_01/parse.py (hunk 8 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: Has been fixed.

File: kalebswartz7/excercise_01/parse.py (hunk 9 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: Added.

File: kalebswartz7/excercise_01/parse.py (hunk 7 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: This has been fixed.

File: kalebswartz7/excercise_01/parse.py (hunk 10 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: This has been changed.

File: kalebswartz7/excercise_01/parse.py (hunk 11 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: They are gone.

File: kalebswartz7/excercise_01/parse.py (hunk 5 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: Docstring added

File: kalebswartz7/excercise_01/parse.py (hunk 12 above)
Comment (kalebswartz7) [reply kalebswartz7/excercise_01/parse.py]: Has been fixed

**Review by mrakitin — APPROVED** [review]
Looks good to me.

**Pull request merged**